cdc/owner/ddl_puller.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package owner
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	timodel "github.com/pingcap/parser/model"
+	"github.com/pingcap/ticdc/cdc/entry"
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/cdc/puller"
+	"github.com/pingcap/ticdc/pkg/config"
+	cdcContext "github.com/pingcap/ticdc/pkg/context"
+	"github.com/pingcap/ticdc/pkg/filter"
+	"github.com/pingcap/ticdc/pkg/regionspan"
+	"github.com/pingcap/ticdc/pkg/util"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+// DDLPuller is a wrapper of the Puller interface for the owner
+// DDLPuller start a puller and listen to the DDL range, add the received DDL into an internal queue
+type DDLPuller interface {
+	// Run runs the DDLPuller
+	Run(ctx cdcContext.Context) error
+	// FrontDDL returns the first DDL job in the internal queue
+	FrontDDL() (uint64, *timodel.Job)
+	// PopFrontDDL returns and pops the first DDL job in the internal queue
+	PopFrontDDL() (uint64, *timodel.Job)
+	// Close closes the DDLPuller
+	Close()
+}
+
+type ddlPullerImpl struct {
+	puller puller.Puller
+	filter *filter.Filter
+
+	mu             sync.Mutex
+	resolvedTS     uint64
+	pendingDDLJobs []*timodel.Job
+	cancel         context.CancelFunc
+}
+
+func newDDLPuller(ctx cdcContext.Context, startTs uint64) (DDLPuller, error) {
+	pdCli := ctx.GlobalVars().PDClient
+	conf := config.GetGlobalServerConfig()
+	kvStorage := ctx.GlobalVars().KVStorage
```

Comment:
how about mvoe kvStorage to L67, make related code more compact together.

Reply:
resolved